### ISSUE TITLE
default group access settings aren't saved for new docs if their controls aren't displayed 

### ISSUE THREAD

**Issue by JonathanReeve** [boonebgorges/buddypress-docs#421]
It looks like the default group access settings aren't being saved unless those settings are available to the user in the HTML. Yet there could be situations where an admin wants to restrict the user's control over the doc access settings, or just clean up the doc user interface, while maintaining the default access settings for a doc. 

Use case: an admin wants to theme buddypress-docs by adding a custom `edit.php` to his theme. He  removes `div#doc-settings` from `edit.php`, or filters `bp_docs_allow_access_settings`, which removes the access settings controls. However, he discovers that without these settings available to the user, a doc created within a hidden group now has public visibility, i.e. the default group access settings aren't applied.

**Comment by christianwach** [comment]
This one's biting me at the moment. A vanilla install of BP and BP Docs exhibits the behaviour that @dcavins describes. No ordinary member of a hidden group can create a BP Doc and associate it with the hidden group during the create process, although the BP Doc can be retrospectively assigned to the group using the BP Doc's "edit" screen. On the "create" screen, the HTML for "Associated Group" and "Access" is never written to the page.

I'm not exactly sure what the ultimate cause of this is, but when the plugin [bails out of setting any caps when there's no `$doc`](https://github.com/boonebgorges/buddypress-docs/blob/master/includes/caps.php#L63), this causes `current_user_can( 'bp_docs_manage' )` to return `false` in the edit template [here](https://github.com/boonebgorges/buddypress-docs/blob/master/includes/templates/docs/single/edit.php#L77) and [here](https://github.com/boonebgorges/buddypress-docs/blob/master/includes/templates/docs/single/edit.php#L98). I can work around this (though I'm not 100% confident that this won't have unforeseen consequences elsewhere) by filtering `bp_docs_map_meta_caps` like this:

```php
/**
 * Allow members of hidden groups to create and associate BP Docs with that group
 *
 * @param array $caps Capabilities for meta capability
 * @param string $cap Capability name
 * @param int $user_id User id
 * @param mixed $args Arguments passed to map_meta_cap filter
 */
function myprefix_docs_map_meta_caps( $caps, $cap, $user_id, $args ) {

	// check cap, location and user
	if ( $cap = 'bp_docs_manage' AND bp_docs_is_doc_create() AND $user_id ) {

		$group_id = null;
		
		// grabbed from BP_Docs_Groups_Integration::get_access_options()
		if ( isset( $_POST['associated_group_id'] ) ) {
			$group_id = intval( $_POST['associated_group_id'] );
		} else if ( isset( $_GET['associated_group_id'] ) ) {
			$group_id = intval( $_GET['associated_group_id'] );
		} else if ( isset( $_GET['group'] ) ) {
			$maybe_group = BP_Groups_Group::get_id_from_slug( $_GET['group'] );
			if ( $maybe_group ) {
				$group_id = $maybe_group;
			}
		}

		// how did we do?
		if ( $group_id AND groups_is_user_member( $user_id, $group_id ) ) {
			$caps = array( 'exist' );
		}

	}

	return $caps;

}

add_filter( 'bp_docs_map_meta_caps', 'myprefix_docs_map_meta_caps', 20, 4 );
```

The issue remains, however, that this presents the user with the UI and I'd rather hide it from them and set the group and the permissions silently. I can do this with CSS, but... well... I'd rather have hidden inputs written into the page. @boonebgorges would you suggest writing them into the page via one of the actions (eg `bp_docs_before_tags_meta_box`) or can you suggest a better way?

**Comment by dcavins** [comment]
Hi Christian-

Part of the problem you're seeing on the doc create screen was resolved
here: https://github.com/boonebgorges/buddypress-docs/issues/482

The problem was a capability issue, and the fix makes the current user the
author of the dummy doc so that the caps pass under the 'creator' case.

-David

**Comment by dcavins** [comment]
For what it's worth, hiding #doc-associated-group and #doc-settings using CSS on the create screen does associate the doc with the source group using the default settings. But the fix above has to be in place so that those meta boxes exist.

**Comment by christianwach** [comment]
@dcavins Dang it, how did I miss that? Thanks for the link, functionality returns!

I'm not entirely persuaded by the "hide via CSS" route - it's too easy to unhide the UI and mess with the group association and doc permissions. I realise a hidden input can also be changed, but it seems to me that this is a classic case of where hidden inputs *should* be used, i.e. when the dev wants to hide these options from the user. Perhaps there's a case for adding some kind of filter or action to `bp_docs_doc_associated_group_markup()` and `bp_docs_doc_settings_markup()` so that writing the form element markup to the page can be short-circuited and replaced with hidden inputs?

**Comment by boonebgorges** [comment]
I agree that hiding with CSS is not ideal.

Hidden inputs are fine, and we could go with the following kind of solution: always print a set of hidden inputs (or a single combined input) with existing Doc settings. On save, check for the existence of the editable values in `$_POST`. If they're not found, use the existing values as a fallback. This is not too difficult to implement.

It still feels weird. Much cleaner would be: if a value is not found in `$_POST`, don't update it. However, as noted above, we need workarounds for checkboxes.

If anyone has strong feelings about this - ideally strong enough to write a patch/PR :) - speak up.

**Comment by christianwach** [comment]
@boonebgorges I presume you mean `<select>` elements rather than checkboxes as I don't see any of those in `#doc-meta` ;-)

For a BP Doc *create*, it looks to me as though  an empty value in `$_POST` (the "None" `<option>` for Associated Group and the "no parent" `<option>` for Parent) means "do not create this setting". There is no empty option for each of the Access `<select>`s, so missing data should be regarded as meaning "use the defaults".

For the BP Doc *edit* screen, it looks to me as though an empty value in `$_POST` (the "None" `<option>` for Associated Group and the "no parent" `<option>` for Parent) is a signal that the setting should be cleared. There is no empty option for each of the Access `<select>`s, so missing data should be regarded as "no change".

Thoughts?

**Comment by boonebgorges** [comment]
Looking more at this, @dcavins is right that the fix in #482 applies here.

@christianwach - I'm not seeing the distinction between create and edit that you've described here. The logic looks like this:

    $associated_group_id = isset( $_POST['associated_group_id'] ) ? intval( $_POST['associated_group_id'] ) : null;

and then later

    // Add to a group, if necessary
    if ( ! is_null( $associated_group_id ) ) {
         bp_docs_set_associated_group_id( $post_id, $associated_group_id );
    }

If the group dropdown is set no None, then `$_POST['associated_group_id']` is '0', which means that `! is_null( $associated_group_id )`. If `$_POST` contains no group info, group associations will never be set in the first place, but existing ones will not be touched. So the original point by @JonathanReeve is still valid, but it's limited to cases where a custom edit.php template is being used, and that custom template doesn't contain all the expected hidden inputs.

**Comment by christianwach** [comment]
Yes, I think I've been conflating two issues. #482 solved the problem I was seeing, which I incorrectly read as being the same as this one. I'll mull the "missing values" issue further.
